Split vision.py into smaller scripts

rahul-tuli opened the pull request:
Refactor: vision.py into 4 scripts
* train.py for training specific tasks
* lr_analysis.py for lr sensitivity analysis
* pr_analysis.py for Kernel Sparsity (pruning) analysis
* export.py for export a model to onnx and store input/outputs

Added:
- Quality fixes for function arguments and * imports
- Scripts tested locally with example commands from old vision.py
- vision.py modified to point user to new scripts
- updation of README
- updation of example commands in recipes
- NmArgumentParser class for parsing dataclasses into arguments
- Tests to show example usage of the parser
- Support for both hyphenated and non-hyphenated arguments

The scripts are tested locally using the following commands on `imagenette` dataset
```python
python integrations/pytorch/train.py \
    --recipe-path integrations/pytorch/recipes/resnet50-pruned.md \
    --arch-key resnet50 --dataset imagenette --dataset-path ~/datasets/ \
    --train-batch-size 256 --test-batch-size 1024


python integrations/pytorch/export.py \
    --arch-key resnet50 --dataset imagenette --dataset-path ~/datasets/



python integrations/pytorch/pr_sensitivity.py \
    --approximate --arch-key resnet50 --dataset imagenette \
    --dataset-path ~/datasets/


python integrations/pytorch/lr_analysis.py \
    --arch-key mobilenet --dataset imagenette \
    --dataset-path ~/datasets/ --batch-size 2
```

rahul-tuli committed:
Refactor: vision.py into 3 scripts
* train.py for training specific tasks
* analysis.py for pruning and lr sensitivity analysis
* export.py for export a model to onnx and store input/outputs

- Quality fixes for function arguments and * imports pending
- Doctrings and usage commands pending
- Scripts tested locally with example ccmmands from old vision.py
- vision.py pending delete
- updation of README files pending

rahul-tuli committed:
Added:
- Quality fixes for function arguments and * imports
- Doctrings and usage commands
- Scripts tested locally with example commands from old vision.py
- vision.py modified to point user to new script
- updation of README
- updation of example commands in recipes

rahul-tuli committed:
Fixed:
- Merge conflicts with main
- classification.transfer_learn_pruned_quantized.md example command
- classification.transfer_learn_pruned.md example command
- tutorial example commands

lucaslie commented on integrations/pytorch/train.py:
```diff
@@ -0,0 +1,406 @@
+from torch.nn import Module
+from torch.utils.data import DataLoader
+
+from helpers import (
```
I am not 100% familiar with the import structure in `sparseml` but would this import work if one tried to execute the script from another directory? 
Or is that not supported?

rahul-tuli replied:
It should work when executed from a different directory @lucaslie 
```python
import sys
print(sys.path)
```

The train, export, analysis scripts have '/home/rahul/projects/sparseml/integrations/pytorch' in their `sys.path` so python should be able to find `helpers.py`, it would work as long as helpers is not in a separate directory. 

Try running the following command

```bash
python sparseml/integrations/pytorch/analysis.py lr_sensitivity \
    --arch-key mobilenet --dataset imagenette \
    --dataset-path ~/datasets --batch-size 2
```

It takes about ~20secs to run. It works on my machine, let me know if you get an error, will change this to a relative import then if needed.

lucaslie replied:
Sounds good. I will check it before submitting my review.

lucaslie replied:
yes, that works :)

lucaslie commented on integrations/pytorch/export.py:
```diff
@@ -0,0 +1,206 @@
+"""
+#####
+Command help:
+usage: export.py [-h] --arch-key ARCH_KEY [--pretrained PRETRAINED]
```
are those up-to-date? Not sure if the argparser changed much because of the split but might be worth double-checking?

rahul-tuli replied:
Yes, those are up to date, generated fresh using --help

lucaslie reviewed (commented):
I checked over it and left suggestions for a few small changes. But I didn't run it to verify that it's working.

markurtz commented:
Putting a note in here, @rahul-tuli please wait to land this until I have some time to look through given the size of the change and impact on testing. I'll try to get through it over the next few days, thanks

rahul-tuli commented:
Thanks @lucaslie taking a deeper look everywhere for references to vision.py

rahul-tuli committed:
Removed: References to vision.py
-Added links to README.md

rahul-tuli commented on integrations/pytorch/train.py:
```diff
@@ -0,0 +1,406 @@
+--nproc_per_node <NUM GPUs> \
+integrations/pytorch/train.py \
+--use-mixed-precision \
+<VISION.PY TRAIN ARGUMENTS>
```
Resolved! Thanks for catching these, don't know how missed them somehow.

rahul-tuli commented on integrations/pytorch/recipes/mobilenet-pruned_quant.md:
````diff
@@ -148,7 +148,7 @@ The training script can be found at `sparseml/integrations/pytorch/vision.py`.
 ```
````
Resolved!

rahul-tuli commented on integrations/pytorch/recipes/resnet50-imagenette-pruned.md:
```diff
@@ -133,7 +133,7 @@ Alternatively, a full walk-through notebook is located at `sparseml/notebooks/py
 *script command:*
```
Resolved!

rahul-tuli commented on integrations/pytorch/recipes/resnet50-pruned.md:
```diff
@@ -73,7 +73,7 @@ The training script can be found at `sparseml/integrations/pytorch/vision.py`.
 *script command:*
```
Resolved!

rahul-tuli commented on integrations/pytorch/recipes/resnet50-pruned_quant.md:
````diff
@@ -186,7 +186,7 @@ The training script can be found at `sparseml/integrations/pytorch/vision.py`.
 ```
````
resolved!

rahul-tuli commented:
> Putting a note in here, @rahul-tuli please wait to land this until I have some time to look through given the size of the change and impact on testing. I'll try to get through it over the next few days, thanks

Sure @markurtz one thing I wanted your take on is:
My Two train of thoughts
1.  Should the arguments be in the same script where they are needed. Will add code duplicacy but a dev-user won't have to go to a separate file to check/change/add arguments.

2. The current implementation removes code duplicacy but a dev-user might have to go to helpers.py to check/change/add arguments

If 1 is preferred over 2, let me know I have that implementation in another local branch.

rahul-tuli commented on integrations/pytorch/README.md:
```diff
@@ -85,8 +85,9 @@ The following table lays out the root-level files and folders along with a descr
-
+| train.py                     | Utility training script for sparsifying classification and detection models in PyTorch (loads SparseML models).       |
+| export.py                     | Utility script for exporting pytorch model to ONNX file along with sample inputs, outputs and labels.                 |
+|analysis.py|Utility script to Run a kernel sparsity (pruning) or learning rate sensitivity analysis for a desired image classification architecture |
```
Done!

rahul-tuli commented on integrations/pytorch/recipes/mobilenet-pruned.md:
```diff
@@ -100,7 +100,7 @@ The training script can be found at `sparseml/integrations/pytorch/vision.py`.
 *script command:*
```
fixed!

rahul-tuli replied:
done!

lucaslie commented on integrations/pytorch/README.md:
```diff
@@ -80,12 +80,14 @@ The following table lays out the root-level files and folders along with a descr
+| [torchvision_sparsification.py](https://github.com/neuralmagic/sparseml/tree/main/integrations/pytorch/torchvision_sparsification.py) | Example training script for sparsifying PyTorch torchvision classification models.                                    |
+| [train.py](https://github.com/neuralmagic/sparseml/tree/main/integrations/pytorch/train.py)                     | Utility training script for sparsifying classification and detection models in PyTorch (loads SparseML models).       |
+| [export.py](https://github.com/neuralmagic/sparseml/tree/main/integrations/pytorch/export.py)                     | Utility script for exporting pytorch model to ONNX file along with sample inputs, outputs and labels.                 |
+|[analysis.py](https://github.com/neuralmagic/sparseml/tree/main/integrations/pytorch/analysis.py)|Utility script to Run a kernel sparsity (pruning) or learning rate sensitivity analysis for a desired image classification architecture |
```
sorry to be nit-picky about it but I would suggest hyper-referencing relative paths, e.g.:

```
 [notebooks](./integrations/pytorch/notebooks)
```

This way it will auto-resolve to the correct branch depending on the branch you are on and IDEs like VSCode can also link to the right directory instead of going to the main branch on github.com

rahul-tuli replied:
Thanks @lucaslie didn't know about this
![](https://media3.giphy.com/media/xUPGcxpCV81ebKh7Vu/giphy.gif)

nit
`[notebooks](notebooks)`

lucaslie commented:
> > Putting a note in here, @rahul-tuli please wait to land this until I have some time to look through given the size of the change and impact on testing. I'll try to get through it over the next few days, thanks
> 
> Sure @markurtz one thing I wanted your take on is:
> My Two train of thoughts
> 
>     1. Should the arguments be in the same script where they are needed. Will add code duplicacy but a dev-user won't have to go to a separate file to check/change/add arguments.
> 
>     2. The current implementation removes code duplicacy but a dev-user might have to go to helpers.py to check/change/add arguments
> 
> 
> If 1 is preferred over 2, let me know I have that implementation in another local branch.

So I was brain-storming a bit and checking over the current set of functionality in `helpers.py`. I was also looking at pytorch's [vision/references](https://github.com/pytorch/vision/tree/master/references), e.g., the [classification reference](https://github.com/pytorch/vision/tree/master/references/classification).

They definitely seem to copy/paste a lot of code across the different reference implementations. But they also have a very lean and short `utils.py` in most sub-folders solely dedicated to the folder. 

I like this model and I think that we should follow it by using `helpers.py` (or renamed as `utils.py`?) as a way to abstract away irrelevant details but also keep it simple. 

So some things we could do to simplify the dependency on `helpers.py` without copy/pasting all the code: 
* Simplify how to argparsers are created. Maybe just one common set of args plus one script-specific `add_args`
* remove global variables in `helpers.py` that we then import 
* remove the `CURRENT_TASK` in the scripts and the corresponding if, else statements in the helper functions. If it doesn't generalize without if, else let's just keep it separate
* remove the global variable for `LOGGER` in `helpers.py` 

The list is probably not complete but hopefully captures the spirit of what I am trying to say.

On a more general note: I am not a fan of enumerating function imports from a module like `helpers` instead of explicitly importing the entire module and then calling it explicitly like `helpers.add_universal_args`. But if that's the style that we generally adopt, please feel free to ignore my comment.

rahul-tuli commented:
Thanks @lucaslie those are very valid points that you've put forth. Will update my design accordingly. Really appreciate the effort. 😇

bfineran reviewed (commented):
Great work @rahul-tuli! Changes look overall good to me, thanks for the thorough update to the script references.

Two things:

1. make sure to test `train.py` using DDP with multiple GPUs (let's sync offline about this)
2. Since the `add_*_args()` functions have a notion of `task` anyways, what is the reasoning for needing to split them out into multiple functions to be called by each task? If we're going to keep the argparse definition in a separate place, it may be nicer to keep it created in a single function, that way a reader will be able to quickly see which args go to which tasks, rather than having to switch between files.

rahul-tuli commented:
> Great work @rahul-tuli! Changes look overall good to me, thanks for the thorough update to the script references.
> 
> Two things:
> 
> 1. make sure to test `train.py` using DDP with multiple GPUs (let's sync offline about this)
> 2. Since the `add_*_args()` functions have a notion of `task` anyways, what is the reasoning for needing to split them out into multiple functions to be called by each task? If we're going to keep the argparse definition in a separate place, it may be nicer to keep it created in a single function, that way a reader will be able to quickly see which args go to which tasks, rather than having to switch between files.


I like this train of thought. Let's sync again offline about the final design. Wil push out changes accordingly. :)

rahul-tuli committed:
Refactor: Task Specific scripts to new design
* train.py
* analysis.py
* export.py

Additionally: Changes all links in following files to hyper-reference-links
* README.md
* tutorials/classification_sparse_transfer_learning_tutorial.md
* tutorials/sparsifying_pytorch_models_using_recipes.md

lucaslie reviewed (changes requested):
Hi @rahul-tuli,

I did another pass over the PR and I really like the structure now. :) Very intuitive for somebody who sees it literally for the first time lol. 

One general question I had, which is probably not related to this particular PR but might be worth considering: 

it seems that the log/results directories that are automatically created during a run (e.g. `pytorch_vision_train`) are not captured by our `.gitignore` and so it pops up as untracked files. Maybe it would be nice to include them in our `.gitignore`?

rahul-tuli committed:
Refactor:According to review comments

rahul-tuli committed:
Refactor:According to review comments again

bfineran reviewed (approved):
LGTM, make sure to test a simple DDP flow in train.py before landing

markurtz reviewed (changes requested):
This is looking good Rahul. I have a few comments throughout and two general ones:
- Let's remove the passing of namespace into the functions. It makes it easier to code now, but makes it harder to debug and track variable state later. An alternative would be using typed classes for that to make it clear what can and cannot be passed through to each function.
- Please be sure to resolve all comments that have been fixed and are no longer open or in discussion

lucaslie reviewed (changes requested):
just another few, minor comments.

bfineran reviewed (dismissed):
should have caught this before, only train.py needs DDP support

rahul-tuli committed:
Add: NmArgumentParser class for parsing dataclasses into command line arguments,
* tests to show example argument parsing
* support for hyphenated and underscore arguments

Split: analysis.py into two scripts, due to missing subparser support in NmArgumentParser

Tested: A DDP workflow with 1 GPU
Refactor: All 4 scripts to use dataclasses for argument parsing

rahul-tuli commented on integrations/pytorch/analysis.py:
```diff
@@ -0,0 +1,477 @@
+    LOGGER.info("created loss: {}".format(loss))
+
+    # device setup
+    module, device, device_ids = model_to_device(model, args.device)
```
Done!

rahul-tuli commented on integrations/pytorch/utils.py:
```diff
@@ -0,0 +1,573 @@
+    :param loggers: List of loggers to use during training process
+    """
+    # optimizer setup
+    if args.optim == "SGD":
```
Done!

rahul-tuli commented on integrations/pytorch/export.py:
```diff
@@ -0,0 +1,230 @@
+
+    args = parser.parse_args()
+
+    utils.append_ddp_defaults_and_preprocessing_args(args=args)
```
Thanks! changed

rahul-tuli commented on integrations/pytorch/analysis.py:
```diff
@@ -0,0 +1,472 @@
+
+
+def lr_sensitivity(
+    args: argparse.Namespace,
```
Changed!

rahul-tuli commented on integrations/pytorch/analysis.py:
```diff
@@ -0,0 +1,472 @@
+
+    if args_.command == utils.Tasks.LR_SENSITIVITY.value:
+        lr_sensitivity(args_, model, train_loader, save_dir)
+    else:
```
Pulled out to a separate script

rahul-tuli commented on integrations/pytorch/train.py:
```diff
@@ -0,0 +1,503 @@
+    parser.add_argument(
+        "--logs-dir",
+        type=str,
+        default=os.path.join("pytorch_vision_train", "tensorboard-logs"),
```
Added

rahul-tuli committed:
Update gitignore

Added:newline

rahul-tuli commented on integrations/pytorch/analysis.py:
```diff
@@ -0,0 +1,472 @@
+        (lr_sensitivity_parser, utils.Tasks.LR_SENSITIVITY),
+    ]
+
+    for _parser, subtask in subtasks:
```
No more for loop

rahul-tuli committed:
Merge branch 'main' into feat-split-vision.py

markurtz committed:
Merge branch 'main' into feat-split-vision.py

lucaslie reviewed (approved):
looks great. Great job!

Final comment: do we need to add Huggingface's licence to the code since our `NmArgumentParser` very much resembles their argparser class or is the current reference `# Inspired from ...` enough?

rahul-tuli commented:
> looks great. Great job!
> 
> Final comment: do we need to add Huggingface's licence to the code since our `NmArgumentParser` very much resembles their argparser class or is the current reference `# Inspired from ...` enough?

We should be good cause we already refer to their license in our notice [here](https://github.com/neuralmagic/sparseml/blob/fe5d49cc2f116b29f25d5ef10c2ae048c13d3b82/NOTICE#L37)

rahul-tuli commented:
> should have caught this before, only train.py needs DDP support

Dismissing review cause it's stale!
